homeassistant/components/broadlink/switch.py

```diff
@@ -54,288 +32,234 @@
+        self._state = False
 
     @property
     def name(self):
         """Return the name of the switch."""
-        return self._name
+        return f"{self._device.name} Switch"
 
     @property
     def assumed_state(self):
-        """Return true if unable to access real state of entity."""
+        """Return True if unable to access real state of the switch."""
         return True
 
     @property
     def available(self):
-        """Return True if entity is available."""
-        return not self.should_poll or self.device.available
+        """Return True if the switch is available."""
+        return self._coordinator.last_update_success
+
+    @property
+    def is_on(self):
+        """Return True if the switch is on."""
+        return self._state
 
     @property
     def should_poll(self):
-        """Return the polling state."""
+        """Return True if the switch has to be polled for state."""
         return False
 
     @property
-    def is_on(self):
-        """Return true if device is on."""
-        return self._state
+    def device_info(self):
+        """Return device info."""
+        return {"identifiers": {(DOMAIN, self._device.unique_id)}}
+
+    @callback
+    def update_data(self):
+        """Update data."""
+        self.async_write_ha_state()
+
+    async def async_added_to_hass(self):
+        """Call when the switch is added to hass."""
+        state = await self.async_get_last_state()
+        self._state = state.state == STATE_ON if state is not None else False
+        self.async_on_remove(self._coordinator.async_add_listener(self.update_data))
 
     async def async_update(self):
-        """Update the state of the device."""
-        if not self.available:
-            await self.device.async_connect()
+        """Update the switch."""
+        await self._coordinator.async_request_refresh()
 
     async def async_turn_on(self, **kwargs):
-        """Turn the device on."""
+        """Turn on the switch."""
         if await self._async_send_packet(self._command_on):
             self._state = True
             self.async_write_ha_state()
 
     async def async_turn_off(self, **kwargs):
-        """Turn the device off."""
+        """Turn off the switch."""
         if await self._async_send_packet(self._command_off):
             self._state = False
             self.async_write_ha_state()
 
+    @abstractmethod
     async def _async_send_packet(self, packet):
-        """Send packet to device."""
-        if packet is None:
-            _LOGGER.debug("Empty packet")
-            return True
+        """Send a packet to the device."""
+
+
+class BroadlinkRMSwitch(BroadlinkSwitch):
+    """Representation of a Broadlink RM switch."""
+
+    def __init__(self, device, name, config):
+        """Initialize the switch."""
+        super().__init__(
+            device, config.get(CONF_COMMAND_ON), config.get(CONF_COMMAND_OFF)
+        )
+        self.entity_id = f"{SWITCH_DOMAIN}.{slugify(name)}"
```

Comment:
This switch is defined by the user in the configuration file. It is associated with the device, but is not registered because it does not have an unique ID, so the user is free to update his configuration file without worrying about the entity registry.

Reply:
Is it impossible to set a unique id for it? We shouldn't avoid the entity registry to cater for the config yaml.

Reply:
We know it's Mac from the discover process. But we'd need to separate the manually added switches. Maybe suffixing name with Mac.

Reply:
In the longer run we want to remove the manual yaml config for this and use an options flow to configure these switches.

We shouldn't use the name in the unique_id. The on/off packet could be used, but those are currently optional in the config schema. That's also weird.

If we would use an options flow we could generate a UUID for each added switch and use that together with the mac address.

Reply:
I think configuring custom switches via options flow would be the ideal scenario. But until we have that, I think we better not register these switches.

Suppose a user wants to delete the switch. He will remove it from the config file hoping it will disappear. But it will remain in the registry, causing frustration. Or a user modifies a switch in the config file. Now he will have two switches in the registry, one before and one after the modification. Not to mention the problems of future migrations when we decide to implement the options flow.

Reply:
My suggestion is to start with this foundational PR and migrate the switches when the options flow is ready, importing the configuration once and for all and storing the switches in JSON files, like those on the remote platform.

Reply:
We can still let the name be the template for the entity_id, right?

Reply:
Sure, but this will bring changes to the configuration schema of the switch platform. Users will need to change their configuration files from:

```yaml
- platform: broadlink
  mac: 34-EA-34-41-3A-3A
  switches:
    tv_sony:
      friendly_name: "Sony TV"
      command_on: 'JgAcAB0dHB44HhweGx4cHR06HB0cHhwdHB8bHhwADQUAAAAAAAAAAAAAAAA='
      command_off: 'JgAaABweOR4bHhwdHB4dHRw6HhsdHR0dOTocAA0FAAAAAAAAAAAAAAAAAAA='
    tv_lg:
      friendly_name: "LG TV"
      command_on: 'JgAcAB0dHB44HhweGx4cHR06HB0cHhwdHB8bHhwADQUAAAAAAAAAAAAAAAA='
      command_off: 'JgAaABweOR4bHhwdHB4dHRw6HhsdHR0dOTocAA0FAAAAAAAAAAAAAAAAAAA='
```

To:
```yaml
switch:
  - platform: broadlink
    mac: 34-EA-34-41-3A-3A
    switches:
      - name: "Sony TV"
        command_on: 'JgAcAB0dHB44HhweGx4cHR06HB0cHhwdHB8bHhwADQUAAAAAAAAAAAAAAAA='
        command_off: 'JgAaABweOR4bHhwdHB4dHRw6HhsdHR0dOTocAA0FAAAAAAAAAAAAAAAAAAA='
      - name: "LG TV"
        command_on: 'JgAcAB0dHB44HhweGx4cHR06HB0cHhwdHB8bHhwADQUAAAAAAAAAAAAAAAA='
        command_off: 'JgAaABweOR4bHhwdHB4dHRw6HhsdHR0dOTocAA0FAAAAAAAAAAAAAAAAAAA='
```
I think it is better, as it facilitates future migrations. Should I proceed?

Reply:
Done.

Reply:
Is there a place where we can display the MAC of the device? I think it would help the configuration of these switches.